### ISSUE TITLE
Javascript: negative int32 values were decoded incorrectly

### ISSUE THREAD

**Issue by stanley-cheung** [google/protobuf#2659]
I discovered this when trying to serialize a protobuf from PHP and deserialize it in JS.

Test case (everything is in version 3.2.0):

`test.proto`:
```
syntax = "proto3";

package math;

message Point {
  int32 latitude = 1;
  int32 longitude = 2;
}
```

`composer.json`:
```
{
  "require": {
    "google/protobuf": "v3.2.0-alpha-1"
  },
  "autoload": {
    "psr-4": {
      "": "./"
    }
  }
}
```

`test.php`:
```
<?php

include_once "vendor/autoload.php";

$p = new Math\Point();
$p->setLatitude(409146138);
$p->setLongitude(-746188906);

$b = $p->encode();

print(bin2hex($b));
```

`package.json`:
```
{
  "dependencies": {
    "google-protobuf": "3.2.0-rc.2"
  }
}
```

`test.js`
```
var messages = require('./test_pb');

var data = '';
process.stdin.setEncoding('utf-8');
process.stdin.on('readable', function() {
    var chunk;
    while (chunk = process.stdin.read()) {
        data += chunk;
    }
});

process.stdin.on('end', function () {
    console.log(data);
    var b = new Uint8Array(new Buffer(data, "hex"));
    var p2 = messages.Point.deserializeBinary(b);

    console.log(p2.getLatitude());
    console.log(p2.getLongitude());
});
```

JS messages were generated by: `protoc --js_out=import_style=commonjs,binary:. test.proto`
PHP messages were generated by: `protoc --proto_path=. --php_out=. ./test.proto`

Run the test:

```
$ php test.php | node test.js
089aa68cc30110969f989c0d
409146138
3548778390
```

Basically, in PHP, a proto is created with 2 int32 fields. One field is set to 409146138. Another to -746188906. The proto is serialized into binary in PHP. I converted the binary into hex and print it to stdout. Pipe the output via stdin into JS. Convert the hex into Uint8Array, then call `.deserializeBinary()`. The proto object I got back in JS shouldn't give me back the number "3548778390" for an "int32" field

**Comment by stanley-cheung** [comment]
@murgatroid99 @TeBoring FYI

**Comment by updogliu** [comment]
FYI, I tried out parsing the varint32 representation of -746188906, i.e. [0x96, 0x9f, 0x98, 0x9c, 0xfd, 0xff, 0xff, 0xff, 0xff, 0x01], using [jspb.BinaryDecoder.prototype.readSignedVarint32](https://github.com/google/protobuf/blob/master/js/binary/decoder.js#L617), and got -746188906 back with no problem.

**Comment by stanley-cheung** [comment]
The JS generated code is decoding and encoding using the readInt32 and writeInt32 functions, which [seems](https://github.com/google/protobuf/blob/3.2.x/js/binary/decoder.js#L857) to be reading and writing fixed bytes

```
proto.math.Point.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break;
    }
    var field = reader.getFieldNumber();
    switch (field) {
    case 1:
      var value = /** @type {number} */ (reader.readInt32());
      msg.setLatitude(value);
      break;
    case 2:
      var value = /** @type {number} */ (reader.readInt32());
      msg.setLongitude(value);
      break;
    default:
      reader.skipField();
      break;
    }
  }
  return msg;
};
```

**Comment by updogliu** [comment]
[Reader.readInt32](https://github.com/google/protobuf/blob/3.2.x/js/binary/reader.js#L608) calls decoder.readSignedVarint32

**Comment by stanley-cheung** [comment]
This is the Uint8Array being received in JS:
```
Uint8Array {
  '0': 8,
  '1': 154,
  '2': 166,
  '3': 140,
  '4': 195,
  '5': 1,
  '6': 16,
  '7': 150,
  '8': 159,
  '9': 152,
  '10': 156,
  '11': 13 }
```

**Comment by updogliu** [comment]
Per the [spec](https://developers.google.com/protocol-buffers/docs/encoding), negative int32 or int64 is always ten bytes long. So seems that PHP encoding is buggy, as it serializes int32 value -746188906 as [96 9f 98 9c 0d], which is only 5 bytes long.

**Comment by stanley-cheung** [comment]
Now I am back to believing that there may be a problem with PHP encoding.

So first I modified my PHP program to print the binary encoding to stdout
```
$b = $p->encode();
print($b);
```

So here's the hex representation of the PHP output:
```
$ php test.php | xxd -l 16 -p
089aa68cc30110969f989c0d
```

Here's the text representation of that
```
$ php test.php | protoc --decode=math.Point test.proto
latitude: 409146138
longitude: -746188906
```

Now I try to encode that text representation again, and convert it to hex, I got something different
```
$ php test.php | protoc --decode=math.Point test.proto | protoc --encode=math.Point test.proto | xxd -l 16 -p
089aa68cc30110969f989cfdffffffff
```

**Comment by TeBoring** [comment]
When encoding negative int32, php encoder doesn't padded it to make it compatible with int64. Working on the fix now.
However, js decoder should be compatible enough to recognize unpadded negative int32 value.

**Comment by updogliu** [comment]
Yeah, make sense. I will fix the JS side.

**Comment by TeBoring** [comment]
Fixed in #2660
